### ISSUE TITLE
Ansible on Windows and mapping network drives

### ISSUE THREAD

**Issue by automatictester** [ansible/ansible#65737]
##### SUMMARY

There is a huge amount of confusion among Ansible users with regards to how to map a Windows drive so that it's accessible to subsequent Ansible tasks. The overall understanding is that it cannot be done - the best you can achieve is that Ansible will create a mapped drive but it will be only visible for interactive sessions, or the other option - you will have to use UNC paths instead of a drive letter, which is not the same thing, really. At least that's the impression I'm getting from reading [this](https://github.com/ansible/ansible/issues/25196), [this](https://github.com/ansible/ansible/issues/33929), [this](https://github.com/ansible/ansible/issues/46713) and [this](https://docs.ansible.com/ansible/latest/modules/win_mapped_drive_module.html#win-mapped-drive-module). I must admit I'm not a Windows expert, though. Long story short, I don't think this is true that you can't use a mapped drive in Ansible tasks, unless my case is very specific.

```
- name: Make sure NFS client is installed (this will also install mount binary)
  win_feature: 
    name: NFS-Client
- name: Copy BAT file
  win_copy:
    src: nfs_mount_script.bat
    dest: C:\nfs_mount_script.bat
- name: Create scheduled task which will mount the network drive
  win_scheduled_task:
    name: nfs_mount
    description: Map NFS share so that it is visible for Ansible tasks
    actions: 
      - path: C:\nfs_mount_script.bat
    triggers:
      - type: boot
    username: SYSTEM
    run_level: highest
- name: Reboot Windows server
  win_reboot:
```

After this, all subsequent Ansible task types I used, be it **win_copy** or others, had no problems using the mounted drive. In the **nfs_mount_script.bat** batch script I used **mount** command, which is NFS specific. I didn't try that out with **net use**, as **mount** simply did the job.

I think it would be useful for many people struggling with a similar problem if it would be explicitly described somewhere in the docs how to mount a network drive so that it can be used by subsequent Ansible tasks.

##### ISSUE TYPE
- Documentation Report

##### ANSIBLE VERSION

```paste below
2.9.1
```

##### CONFIGURATION

```paste below
N/A
```

##### OS / ENVIRONMENT

Ansible controller:
Centos, Python 3

Ansible host:
Windows Server 2016

##### ADDITIONAL INFORMATION

Communication:
WinRM, CredSSP, HTTPS listener

**Comment by automatictester** [comment]
@jborean93

**Comment by jborean93** [comment]
This is mostly what I was referring to in my comment at https://github.com/ansible/ansible/issues/46713#issuecomment-440930802

> This is still mapped per user, technically you can create a global mount but it has a few caveat, see my blog post for more details

I will get into the what I meant by caveats in the reply but first I'll break down the mapped drive implementation in Windows first. A mapped drive is stored as a [DosDevice](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/introduction-to-ms-dos-device-names) and a DosDevice can either be a [local or global name](https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/local-and-global-ms-dos-device-names). The global DosDevices are visible system wide whereas local DosDevices are only visible to the logon session they are created in (local also trumps global in the event of a conflict). When creating a mapped drive, Windows creates a DosDevice in the local DosDevices namespace so traditionally they are only visible to that logon session. These devices are removed once the logon session is finished (local) or when the host is rebooted (global).

There are 3 scenarios that slightly alter this behaviour:

1. When you persist a network drive, the settings are stored in the registry and on an actual interactive (RDP or console) logon Windows will recreate the DosDevice for the current logon session
2. When the [EnableLinkedConnections](https://support.microsoft.com/en-us/help/3035277/mapped-drives-are-not-available-from-an-elevated-prompt-when-uac-is-co) registry flag is applied, any DosDevices created for the limited user logon is also applied to the linked admin logon session
3. The SYSTEM account has no local DosDevices namespace, any drives that are mounted as the SYSTEM account are created as a global DosDevice that is visible to all

A WinRM logon is a network logon and there is no mechanism in Windows to automatically recreate those DosDevices stored in the registry. This is why typically cannot access mapped drives in an Ansible task. The only exception to this rule (that I know off) is scenario 3 I mentioned above. Because the mapped drive is mapped to the global DosDevices namespace it will be visible to all logon sessions and not just the one that created it. This is the behaviour you are seeing right now where running the scheduled task as SYSTEM will create the global mapping. You could also do the following with win_mapped_drive to achieve the same thing:

```
- win_mapped_drive:
    letter: Z
    path: \\server\nfs\mount
    state: present
  become: yes
  become_user: SYSTEM
```

This makes the drive visible for all accounts on the Windows host, even through WinRM, but here are the problems I was referring to:

1. On a reboot, the mapping settings in the registry at `HKCU:\Network\<drive>` (on the SYSTEM's hive) is not automatically mapped to a DOS Device as the SYSTEM account never goes through an interactive logon
2. When you rerun the win_mapped_drive task after a reboot, Windows still thinks the user has a mapped drive at Z and won't remap it
    * This is because the SYSTEM's user hive at `HKCU:\Network\<drive>` still exists (Windows views this as a valid mapping entry) it's just never been mounted to the global DosDevices (point 1)
    * You can prove this by running `net use` under the SYSTEM account after a reboot to see the drive is still there (just not mapped)
3. This is visible to all accounts on the host, if you want it viewable in WinRM for only 1 user account then you are out of luck
4. Less of a problem for NFS but SMB typically requires authentication and in WinRM this isn't available by default. You would have to use CredSSP, Kerberos with credential delegation, or become to access the files
    * This is also true for accessing the path using the UNC path anyway but having a global mount does not make this go away
    * I'm not aware of any way to store a credential for network location globally so mapped drives won't fix this for you
5. Explorer does not play nicely with global DosDevices of a mapped drive, they can appear as disconnected yet still be accesible which is not great for end users

To prove point number 2, run the following tasks

```
- win_mapped_drive:
    letter: Z
    path: \\192.168.56.1\home\nfsshare
    state: present
  become: yes
  become_user: SYSTEM

- win_shell: |
    net use *
    Get-PSDrive -Name Z
  become: yes
  become_user: SYSTEM

- win_reboot:

- win_shell: |
    net use *
    Get-PSDrive -Name Z -ErrorAction Ignore
  become: yes
  become_user: SYSTEM
```

```
TASK [win_mapped_drive] *******************************************************************************************************************************************************************************************
changed: [win-adhoc] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "letter": "Z",
            "password": null,
            "path": "\\\\192.168.56.1\\home\\nfsshare",
            "state": "present",
            "username": null
        }
    }
}

TASK [win_shell] **************************************************************************************************************************************************************************************************
changed: [win-adhoc] => {
    "changed": true,
    "cmd": "net use\nGet-PSDrive -Name Z",
    "delta": "0:00:01.078153",
    "end": "2019-12-11 09:05:07.502055",
    "rc": 0,
    "start": "2019-12-11 09:05:06.423902",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "New connections will be remembered.\r\n\r\n\r\nStatus       Local     Remote                    Network\r\n\r\n-------------------------------------------------------------------------------\r\n             Z:        \\\\192.168.56.1\\home\\nfsshare \r\n                                                NFS Network\r\nThe command completed successfully.\r\n\r\n\r\nName           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation\r\n----           ---------     --------- --------      ----                                               ---------------\r\nZ                  84.68        767.26 FileSystem    \\\\192.168.56.1\\home\\nfsshare                                      \r\n\r\n\r\n",
    "stdout_lines": [
        "New connections will be remembered.",
        "",
        "",
        "Status       Local     Remote                    Network",
        "",
        "-------------------------------------------------------------------------------",
        "             Z:        \\\\192.168.56.1\\home\\nfsshare ",
        "                                                NFS Network",
        "The command completed successfully.",
        "",
        "",
        "Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation",
        "----           ---------     --------- --------      ----                                               ---------------",
        "Z                  84.68        767.26 FileSystem    \\\\192.168.56.1\\home\\nfsshare                                      ",
        "",
        ""
    ]
}

TASK [win_reboot] *************************************************************************************************************************************************************************************************
changed: [win-adhoc] => {
    "changed": true,
    "elapsed": 42,
    "rebooted": true
}

TASK [win_shell] **************************************************************************************************************************************************************************************************
fatal: [win-adhoc]: FAILED! => {
    "changed": true,
    "cmd": "net use\nGet-PSDrive -Name Z -ErrorAction Ignore",
    "delta": "0:00:01.093770",
    "end": "2019-12-11 09:05:56.281733",
    "msg": "non-zero return code",
    "rc": 1,
    "start": "2019-12-11 09:05:55.187962",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "New connections will be remembered.\r\n\r\n\r\nStatus       Local     Remote                    Network\r\n\r\n-------------------------------------------------------------------------------\r\nUnavailable  Z:        \\\\192.168.56.1\\home\\nfsshare \r\n                                                NFS Network\r\nThe command completed successfully.\r\n\r\n",
    "stdout_lines": [
        "New connections will be remembered.",
        "",
        "",
        "Status       Local     Remote                    Network",
        "",
        "-------------------------------------------------------------------------------",
        "Unavailable  Z:        \\\\192.168.56.1\\home\\nfsshare ",
        "                                                NFS Network",
        "The command completed successfully.",
        ""
    ]
}
```

We can see that after the reboot the `Z` mapped drive is not longer accessible (mounted) but `net use` under the SYSTEM account still sees `Z` as a valid entry which is why win_mapped_drive won't re-add it after a reboot.

There is a way you can set Windows to mount the drive globally on a reboot which requires the following win_regedit task.

```
- name: create a global mapped drive with persistence
  win_regedit:
    path: HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\DOS Devices
    name: 'Z:'
    data: '\Device\MRxNfs\;Z:00000000000003e7\server\share'
    type: string
    state: present
```

_Note: the device name for an SMB target won't be `MRxNfs` but `LanmanRedirector`._

This particular registry key is read by the Windows startup process and will create the global mapping of the drive that is viewable in all logon sessions which does what you want. This may be similar to what your batch file does to make it persist upon a reboot but honestly that's just a guess as I can't see the script itself. There might be another mechanism that I am unaware about but after researching this process I could never find anything else.

Due to the weird caveats and limitations with a globally mapped network drive our stance is to not document it at all as it just confused people too much. The recommendation we have always stated is to use the full UNC path (UNC also works for NFS mounts and not just SMB) when accessing network locations as this is always available and does not rely on the host to be set up in a particular way. The whole credential delegation thing is still an issue with UNC paths but mapped drives are also affected by that issue so it's not like they add anything useful. The only time you should use the win_mapped_drive module is if you want to create a mapped drive config entry for a user to use interactively. It should never be used as part of an actual Ansible task.

TLDR: Yes it is technically possible but still not something we want to recommend or even document due to the numerous caveats and limitations that apply to it (see the length of this reply). Just use the UNC path and all will be right in the world.

PS: I tried to cover all this in a blog post that you might be useful https://www.bloggingforlogging.com/2018/11/22/windows-mapped-drives-what-the-hell-is-going-on/.
